contracts/tasks/storageSlots.js

```diff
@@ -0,0 +1,156 @@
+const path = require('path');
+const { promises, existsSync, mkdirSync } = require('fs');
+
+const {
+  getStorageLayout,
+  getVersion,
+  getUnlinkedBytecode,
+  Manifest,
+  getStorageLayoutForAddress,
+  isCurrentValidationData,
+  assertStorageUpgradeSafe
+} = require('@openzeppelin/upgrades-core');
+
+const getStorageFileLocation = (hre, contractName) => {
+  const isFork = process.env.FORK === "true";
+  const isLocalhost = !isFork && hre.network.name === "localhost";
+  const isRinkeby = hre.network.name === "rinkeby";
+  const isMainnet = hre.network.name === "mainnet";
+
+  let folder = 'localhost';
+  if (isFork || isMainnet) {
+    folder = 'mainnet';
+  } else if (isRinkeby) {
+    folder = 'rinkeby';
+  }
+  
+  const layoutFolder = `./storageLayout/${folder}/`
+  if (!existsSync(layoutFolder)){
+    mkdirSync(layoutFolder);
+  }
+
+  return `${layoutFolder}${contractName}.json`;
+}
+
+const getStorageLayoutForContract = async (hre, contractName) => {
+  const { provider } = hre.network;
+  const manifest = await Manifest.forNetwork(provider);
+  const validations = await readValidations(hre);
+  const implFactory = await hre.ethers.getContractFactory(contractName);
+  const unlinkedBytecode = getUnlinkedBytecode(validations, implFactory.bytecode);
+  const version = getVersion(unlinkedBytecode, implFactory.bytecode);
+
+  return getStorageLayout(validations, version);
+}
+
+const loadPreviousStorageLayoutForContract = async (hre, contractName) => {
+  const location = getStorageFileLocation(hre, contractName);
+
+  // new contract
+  if (!existsSync(location)) {
+    return null;
+  }
+
+  return JSON.parse(await promises.readFile(location, 'utf8'));
+}
+
+const storeStorageLayoutForContract = async (hre, contractName) => {
+  const layout = await getStorageLayoutForContract(hre, contractName);
+  const storageLayoutFile = getStorageFileLocation(hre, contractName);
+
+  // pretty print storage layout for the contract
+  await promises.writeFile(storageLayoutFile, JSON.stringify(layout, null, 2));
+  console.log(`Storage slots layout for ${contractName} saved to ${storageLayoutFile} `);
+}
+
+const getAllEligibleContractNames = async (hre) => {
+  const contractNames = Object.keys(await hre.deployments.all());
+
+  const excludeContracts = [
+    'CurveUSDCStrategy',
+    'CurveUSDTStrategy',
+    'MinuteTimelock',
+    'OpenUniswapOracle',
+    'RebaseHooks'
+  ];
+
+  // TODO: check that excluding contracts below + proxy ones is OK
```

Comment:
Proxies are not upgradeable so make sense to exclude them.

Reply:
👍